[1, 2단계 - 체스] 아톰(이하늘) 미션 제출합니다.

le2sky opened the pull request:
조앤, 안녕하세요! 아톰입니다. 
페어인 위브와 체스 1,2 단계를 구현했습니다.
리뷰 잘 부탁드립니다. 감사합니다. 🙇🏻‍♂️

le2sky committed:
docs: 구현 기능 목록 작성

le2sky committed:
feat: 체스판 행열 생성 기능 구현

le2sky committed:
feat: 체스판 좌표 생성 기능 구현

le2sky committed:
docs: 문서 수정

- 누락된 요구사항을 추가 작성했다.
- 구현 기능을 추가 작성했다.
- 요구사항을 정리했다.

le2sky committed:
feat: 진영 생성 기능 구현

le2sky committed:
feat: 기물 종류 생성 기능 구현

le2sky committed:
feat: 기물 생성 기능 구현

le2sky committed:
feat: 체스판 생성 기능 구현

le2sky committed:
feat: 체스판 출력 로직 구현

le2sky committed:
feat: 시작 종료 입력 기능 구현

le2sky committed:
fix: 킹과 퀸 위치 변경

le2sky committed:
refactor: 체스 기물 출력 로직 개선

le2sky committed:
docs: 문서 수정

- 구현 기능을 체크했다.
- 누락된 요구사항을 작성했다.
- 킹과 퀸 위치를 정정했다.

le2sky committed:
style: 메서드 선언 개행 추가

le2sky committed:
refactor: Piece를 추상화하도록 개선

le2sky committed:
feat: 나이트 이동 가능 좌표 계산 기능 구현

le2sky committed:
feat: 킹 이동 가능 좌표 계산 기능 구현

le2sky committed:
feat: 룩 이동 가능 좌표 계산 기능 구현

le2sky committed:
feat: 비숍 이동 가능 좌표 계산 기능 구현

le2sky committed:
feat: 퀸 이동 가능 좌표 계산 기능 구현

le2sky committed:
feat: 폰 이동 가능 좌표 계산 기능 구현

le2sky committed:
refactor: 테스트 클래스 패키지 정리

le2sky committed:
feat: 지정된 기울기 방향 경로 생성 기능 구현

le2sky committed:
fix: 기물 이동 가능 좌표 계산 기능 개선

- 불필요한 방향 경로를 제거하도록 개선
- Direction을 사용하도록 개선

le2sky committed:
refactor: 도메인 패키지 분리

le2sky committed:
feat: 체스 기물 이동 기능 구현

le2sky committed:
docs: 문서 수정

- 누락된 기능을 추가 작성했다.
- 신규 도메인 용어를 추가했다.
- 구현 기능을 체크했다.

le2sky committed:
feat: 이동 방향 입력 기능 구현

le2sky committed:
docs: 문서 수정

- 구현 기능을 체크했다.
- 누락된 기능을 추가 작성했다.

le2sky committed:
refactor: 초기 체스판 생성 로직 분리

le2sky committed:
refactor: 체스판 코드 개선

- 메서드를 분리했다.
- 불필요한 메서드를 제거했다.

le2sky committed:
refactor: 상수 접근 제어자 및 선언 순서 개선

le2sky committed:
fix: 폰 초기 이동 로직 버그 개선

seovalue commented on docs/README.md:
```diff
@@ -0,0 +1,96 @@
+## 도메인 용어 정리
```
👍 오 이거 넘 좋네요

seovalue commented on docs/README.md:
```diff
@@ -0,0 +1,96 @@
+        - [x] source 좌표에 기물이 없으면 예외를 발생한다.
+        - [x] target 좌표에 기물이 있다면, 이동할 수 없다.
+        - [x] source 좌표에 있는 기물이 target 좌표로 이동 가능한 경로가 존재하지 않을 경우, 이동할 수 없다.
+        - [x] target 으로 가는 경로에 기물이 존재하면, 이동할 수 없다.
```
Knight의 경우엔 기물을 뛰어넘을 수 있는 걸로 알아요!
https://www.chess.com/ko/terms/chess-knight-ko

le2sky replied:
감사합니다! 기물의 움직임을 더욱 신경쓰도록 해보겠습니다. 🙇🏻‍♂️

seovalue commented on src/main/java/chess/Controller.java:
```diff
@@ -0,0 +1,69 @@
+import chess.view.InputView;
+import chess.view.OutputView;
+
+class Controller {
```
Controller인 이유가 무엇인가요?

le2sky replied:
제가 아직 많이 부족해서 질문의 의도를 정확히 파악하기 어려운 것 같아요. 😥
해당 클래스는 view를 업데이트하며, view에서 전달받은 데이터를 이용해 도메인 객체를 조작합니다.
따라서, Controller라고 생각했어요. 

아무래도 제가 놓치고 있는 부분이 있는 것 같아요. 추가 피드백 남겨 주시면, 제 성장에 많은 도움이 될 것 같아요. 감사합니다. 🙇🏻‍♂️

seovalue replied:
의도는 따로 없고, 말 그대로 Controller로 지어주신 이유가 궁금했어요.
지금도 여전히 Controller라고 명명되어있는데, 단순히 Controller라고 지었을 때 사용하는 쪽에서 이 컨트롤러가 정확히 어떤 역할을 하는 것인지 알 수 있을까요?ㅎㅎ 컨트롤러가 무엇일까요?

le2sky replied:
Controller 클래스가 하는 일은 체스 게임의 흐름을 제어합니다. chess 패키지의 단일 컨트롤러이기 때문에, 그 의도가 잘 전달될 것이라고 생각했어요. 하지만, 조앤에게 설명을 해야하니 제 의도가 제대로 전달되지 않은 것 같아요. 

현재는 ChessController라는 이름으로 변경해봤어요. 조앤이 보시기에 적절한 반영일까요?

seovalue commented on src/main/java/chess/Controller.java:
```diff
@@ -0,0 +1,69 @@
+
+    private void playChess(Board board) {
+        outputView.printBoard(board);
+        String read = inputView.readCommand();
```
read라는 변수명에 대해서 어떻게 생각하세요

le2sky replied:
해당 변수명으로 다른 개발자에게 의도를 명확히 전달하기 어려울 것 같아요. 🥲 
전반적으로 더욱 의도를 잘 전달할 수 있는 이름을 짓도록 노력해보겠습니다. 감사합니다.  🙇🏻‍♂️

seovalue commented on src/main/java/chess/Controller.java:
```diff
@@ -0,0 +1,69 @@
+    private void playChess(Board board) {
+        outputView.printBoard(board);
+        String read = inputView.readCommand();
+        if (read.startsWith("move ")) {
```
`move ` 와 같은 것은 띄어쓰기를 기준으로 분리해서 {ex. move, a1, a2} 사용해볼까요?

seovalue commented on src/main/java/chess/Controller.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    private List<Coordinate> parseToCoordinates(String read) {
+        String[] moveCommand = read.split("move ");
```
> 배열 대신 컬렉션을 사용한다.

seovalue replied:
다른 배열 부분도 마찬가지예요~

seovalue commented on src/main/java/chess/Controller.java:
```diff
@@ -0,0 +1,69 @@
+            return;
+        }
+
+        throw new IllegalArgumentException("잘못된 입력입니다.");
```
<img width="913" alt="Screenshot 2024-03-21 at 23 00 46" src="https://github.com/woowacourse/java-chess/assets/48412963/c5a3ff8d-4e92-4c7b-9130-2b0202efaf7d">

유저에게 stacktrace를 보여줄 필요가 있을까요?

le2sky replied:
현재 예외를 캐치해서 재시도를 할 수 있도록 개선해봤습니다! 
만약 재시도를 하지 않는다고 해도 사용자에게 stacktrace를 보여줄 필요는 없을 것 같아요. 
피드백 감사합니다. 😀

seovalue commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -0,0 +1,109 @@
+
+    private void validateTargetExist(Coordinate target) {
+        if (pieces.containsKey(target)) {
+            throw new IllegalStateException("목적지 좌표에 기물이 이미 존재합니다.");
```
```
move a2 a4
move b7 b5
move a4 b5
```
위 케이스는 폰이 상대팀의 폰을 잡아야하는 케이스인데, 기물이 이미 존재한다는 에러가 발생하네요~

le2sky replied:
조앤 감사합니다. 현재 기물을 잡는 로직까지 구현했습니다! 🙇🏻‍♂️

seovalue commented on src/main/java/chess/domain/piece/Team.java:
```diff
@@ -0,0 +1,18 @@
+package chess.domain.piece;
```
패키지를 구분하는 기준이 무엇인가요?

le2sky replied:
주로 기능을 중심으로 분류하려고 노력합니다. `board`와 `piece` 패키지가 제공하는 기능이 다르다고 생각했어요.  

`domain.board` 패키지는 기물의 위치와 보드의 크기를 관리하는 영역으로. 
`domain.piece` 패키지는 기물의 움직임을 관리하는 영역으로 나누어봤어요.

`Team` 같은 경우, 진영에 따라서 전진 방향이 다르며, 진영에 따라 상대 기물을 공격하고, 아군 기물을 피할 수 있으니. 기물의 움직임에 
필요한 녀석이라고 생각했어요. 따라서, `domain.piece` 패키지 내부에 존재합니다.

제가 말로 설명하려고 해도 무언가 이상한 부분이 있는 것 같아요. 제가 놓치고 있는 부분이 있을까요?
추가적으로 조앤만의 기준이나, 실무에서 패키지를 구분하는 방법을 알려주시면 제 성장에 큰 도움이 될 것 같아요.

seovalue replied:
여러가지 방법이 있을 것 같아요.
layer 단위로 나누기도 하고요 (controller가 있는 presenter, application, persistence)
말씀주신 것과 유사하게 도메인 단위로 나누기도 하고요 (board, piece..)ㅎㅎ
team을 어떤 이유로 piece 패키지에 두신 건지 이유가 궁금했던거였어요ㅎㅎ

seovalue commented on src/main/java/chess/domain/board/Direction.java:
```diff
@@ -0,0 +1,47 @@
+                .orElseThrow(() -> new IllegalArgumentException("같은 곳으로는 이동할 수 없습니다."));
+    }
+
+    public List<Coordinate> createPath(Coordinate start) {
+        List<Coordinate> coordinates = new ArrayList<>();
+        int fileWeight = entry.getKey();
+        int rankWeight = entry.getValue();
+        int nextRank = start.getRank() + rankWeight;
+        char nextFile = (char) (start.getFile() + fileWeight);
+
+        while (nextRank >= 1 && nextRank <= 8 && nextFile >= 'a' && nextFile <= 'h') {
+            coordinates.add(new Coordinate(nextRank, nextFile));
+            nextRank += rankWeight;
+            nextFile += fileWeight;
+        }
+
+        return coordinates;
```
해당 메서드 메서드 라인 수도 그렇고, 전반적으로 가독성을 높이는 리팩터링을 해볼까요?

seovalue commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -0,0 +1,109 @@
+        validateStuckPath(target, realPath);
+    }
+
+    private List<Coordinate> createRealPath(List<Coordinate> possibleCoordinate) {
```
RealPath가 의미하는게 무엇인가요?

le2sky replied:
다른 기물에 막히지 않고 갈 수 있는 실제 경로를 RealPath라고 지어봤습니다! 
현재는 다른 기물의 방해없이 쭉 미끄러질 수 있는 경로를 뜻하는 SlidingPath로 변경해봤어요. 
의도를 잘 드러내는 이름일까요?

seovalue replied:
ㅋㅋㅋㅋㅋㅋslidingPath 너무 귀여운데요 
하지만 제 생각에는 그냥 `path`로 명명해주는 것이 가장 깔끔하다고 생각해요~

seovalue commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    private void validateCasePawn(Coordinate source, Coordinate target) {
+        if (!(pieces.get(source) instanceof Pawn)) {
```
instanceof를 대체할 순 없을까요?

le2sky replied:
현재 instanceof를 제거하도록 변경해봤습니다. 어색한 부분이 있다면 추가 피드백 부탁드립니다! 😀

seovalue commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -0,0 +1,109 @@
+            return;
+        }
+
+        if (Math.abs(source.getRank() - target.getRank()) == 2 &&
```
조건문이 복잡할 땐 메서드로 분리해볼까요~

seovalue commented on src/main/java/chess/domain/piece/King.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @Override
+    public List<Coordinate> findMovablePath(Coordinate start, Coordinate destination) {
```
> 함수(또는 메서드)의 길이가 10라인을 넘어가지 않도록 구현한다.


전체적으로 기본 원칙을 지켜주세요~

seovalue commented on src/main/java/chess/view/Converter.java:
```diff
@@ -0,0 +1,49 @@
+import chess.domain.piece.Piece;
+import chess.domain.piece.Team;
+
+class Converter {
```
컨버터는 무엇을 converting하는 converter인지, 사용처에서 알 수 있을까요?

le2sky replied:
사용처에서 알기 어려울 것 같아요. 현재는 ViewDataConverter라고 이름을 지어봤어요. 의도를 명확히 전달하는 이름일까요?

seovalue replied:
기존에 Converter보다는 나은 것 같아요! 👍

seovalue commented on src/test/java/chess/domain/board/BoardTest.java:
```diff
@@ -0,0 +1,275 @@
+                .isInstanceOf(NoSuchElementException.class);
+    }
+
+    @DisplayName("target 좌표에 기물이 있으면 예외를 발생한다.")
```
target 좌표에 기물이 있을 때엔, 상대팀의 말을 잡을 수도 있다는 것을 고려해서 다시 구현해주세요~

seovalue reviewed (changes requested):
안녕하세요, 아톰. 조앤이에요ㅎㅎ
체스 미션 잘 수행해주셨네요~! 코멘트 남겼는데 확인부탁드리고, 전체적으로 기본 프로그래밍 요구사항을 지켜보면 좋을 것 같아요ㅎㅎ

le2sky committed:
docs: 코드 개선을 위한 문서 정리

le2sky committed:
refactor: 시그니처 변경을 위한 메서드 선언 추가

le2sky committed:
refactor: 킹 이동 판단 로직 개선

le2sky committed:
refactor: null 대신 널 객체를 반환하도록 개선

le2sky committed:
feat: 디버깅을 위한 toString 재정의

le2sky committed:
feat: 같은 팀 판단 기능 구현

le2sky committed:
docs: 문서 수정

- 구현 기능을 체크했다.
- 누락된 요구사항을 추가 작성했다.

le2sky committed:
chore: 가중치 적용 로직 개선 TODO 작성

le2sky committed:
refactor: 나이트 이동 판단 로직 개선

le2sky committed:
chore: 킹 테스트 데이터 주석 추가

le2sky committed:
refactor: 룩 이동 판단 로직 개선

le2sky committed:
feat: 기물 존재 여부 조회 기능 구현

le2sky committed:
refactor: 비숍 이동 판단 로직 개선

le2sky committed:
refactor: 퀸 이동 판단 로직 개선

le2sky committed:
refactor: 폰 이동 판단 로직 개선

le2sky committed:
feat: 적군 판단 기능 구현

le2sky committed:
docs: 구현 기능 체크

le2sky committed:
refactor: Board 코드 개선

- 불필요한 로직을 제거했다.
- 폰을 내부로 캡슐화했다.
- 누락된 테스트 코드를 추가 작성했다.

le2sky committed:
docs: 문서 수정

- 체스판에 대한 누락된 요구사항을 추가 작성 했다.
- 구현 기능을 체크했다.

le2sky committed:
refactor: 미사용 메서드 제거

le2sky committed:
refactor: direction 패키지 위치 변경

le2sky committed:
fix: 기물 종류 출력 로직 버그 개선

le2sky committed:
fix: 슬라이딩이 가능한 기물들의 버그 개선

- (버그) 중간에 기물은 없지만, 타겟 기물에게 막혀 공격이 불가능했다.
- (개선) 버그를 개선하고, Queen, Rook, Bishop의 공격 테스트를 추가 작성했다.

le2sky committed:
refactor: 중복 코드 추상 클래스로 이동

le2sky committed:
refactor: 슬라이딩 가능 기물들 추상 클래스로 중복 로직 제거

le2sky committed:
refactor: direction 개선

- 신규 방향을 추가했다.
- 기존 방향의 이름을 개선했다.

le2sky committed:
refactor: 슬라이딩 불가능 기물들 추상 클래스로 중복 로직 제거

le2sky committed:
refactor: direction 책임 분산 및 weight 레코드를 사용하도록 개선

le2sky committed:
refactor: 누락된 abstract 키워드 추가 작성

le2sky committed:
style: enum trailing comma 제거

le2sky committed:
feat: 가중치 적용 기능 구현

le2sky committed:
docs: 문서 수정

- 구현 기능을 체크했다.
- 누락된 요구사항을 추가 작성했다.
- 불필요한 요구사항을 제거했다.

le2sky committed:
refactor: 개선된 신규 좌표 생성 로직 적용

le2sky committed:
feat: 가중치 랭크 곱셈 기능 구현

le2sky committed:
refactor: pawn 관련 코드 가독성 개선

le2sky committed:
refactor: 멤버 변수 가시성 변경

le2sky committed:
docs: 문서 수정

- 신규 도메인 용어를 추가했다.
- 구현 기능을 체크했다.
- 신규 요구사항을 추가 작성했다.

le2sky committed:
refactor: 미사용 메서드 제거

le2sky committed:
fix: 폰이 전진으로 적군 기물을 잡을 수 없도록 개선

le2sky committed:
docs: 누락된 요구사항 추가작성

le2sky committed:
feat: 사용자 재입력 로직 구현

le2sky committed:
refactor: 기물 초기 위치 값 팩토리 내부로 이동

le2sky committed:
style: trailing comma 제거

le2sky committed:
refactor: 메서드 선언 순서 개선

le2sky committed:
refactor: 파일과 랭크를 레코드로 변경

le2sky committed:
refactor: 반복되는 예외를 커스텀 예외로 변경

le2sky committed:
refactor: 초기 랭크 판단 로직 팩터리 클래스 내부로 이동

le2sky committed:
refactor: 출력 관련 클래스 네이밍 개선

le2sky committed:
refactor: 뷰, 컨트롤러 코드 개선

- 뷰에서 전달되는 데이터를 Command로 포장했다.
- 컨트롤러 예외 처리 로직을 개선했다.

le2sky committed:
refactor: MoveCommand 관련 로직 개선

- 좌표 생성 로직을 포함하도록 변경했다.
- 가독성을 위해서 인덱스 값을 상수화했다.
- 메서드 명과 변수명들을 다시 지어줬다.

le2sky committed:
refactor: 매개변수 순서 변경

le2sky committed:
refactor: 상수 정리

seovalue commented on docs/README.md:
```diff
@@ -31,23 +32,24 @@
 
 - [x] 좌표를 생성할 수 있다.
     - [x] 좌표는 행,열을 가진다.
+- [x] 가중치를 받아서 새로운 좌표를 생성할 수 있는지 검사한다.
```
가중치가 의미하는 바가 무엇인가요?

le2sky replied:
좌표의 각 Rank와 File에 일정한 값을 더해서, 새로운 좌표를 만들도록 구현했어요. 이때 사용되는 일정한 값을 가중치라고 정해주었습니다. 용도에 맞는 단어 선택일까요?!

le2sky replied:
현재 가중치를 받는다 -> "가중치를 더한다"로 변경해봤어요!

seovalue commented on src/main/java/chess/view/ViewDataConverter.java:
```diff
@@ -8,9 +8,7 @@
-    private static final String DEFAULT_SHAPE = ".";
+class ViewDataConverter {
 
     List<List<String>> convertToViewData(Board board) {
```
컬렉션의 중첩(List가 List를 갖는 형태)을 어떻게 해결할 수 있을까요? 

예를 들어서, Converter.convertToViewData를 했을 때, List<List<String>>이 반환되면 그 안에 어떤 값이 담겨있는지 convertToViewData 내부를 까보지 않고선 알 수 없을 것 같아요.

le2sky replied:
convertToViewData는 말씀주신 것처럼, List<List<String>>을 반환합니다.
이렇게 반환한다면, 사용처에서 해당 뷰 데이터를 어떤식으로 사용해야하는지 모를 것 같아요!
대신 CharSequence를 반환하도록 개선해봤어요! 어색한 부분이 있다면, 추가 피드백 부탁드려요. 🙇🏻‍♂️

seovalue commented on src/main/java/chess/view/ViewDataConverter.java:
```diff
@@ -26,20 +24,12 @@ private List<String> createRowData(Board board, int rankValue) {
-        String shape = PieceShape.valueOf(piece.getType()).getShape();
+    private String convertToViewData(Piece piece) {
+        String shape = PieceShapeSelector.selectShape(piece.getType());
         if (piece.getTeam() == Team.BLACK) {
```
piece.isBlackTeam()

le2sky replied:
piece 인터페이스에 isBlackTeam을 만들면, 팀의 종류가 인터페이스에 스며드는 것이 아닌지 염려됩니다. 🥲

턴 개념을 만들면서 기존에 있던 piece.isSameTeam의 파라미터 타입을 Team으로 변경해봤어요. 이를 사용해서, 

```java
piece.isSameTeam(Team.BLACK);
```
현재는 이런 방식으로 개선해봤어요. 많이 어색한가요?

seovalue commented on src/main/java/chess/domain/piece/AbstractPiece.java:
```diff
@@ -1,21 +1,50 @@
+    private void validateSameTeam(Coordinate target, Board board) {
+        Piece targetPiece = board.findByCoordinate(target);
+        if (isSameTeam(targetPiece)) {
+            throw new IllegalStateException("아군 기물은 공격할 수 없습니다.");
```
<img width="179" alt="Screenshot 2024-03-25 at 23 23 35" src="https://github.com/woowacourse/java-chess/assets/48412963/c9def989-44b6-48a0-b5bd-d2f071963b7b">

turn을 번갈아가면서 게임을 할 수 있도록 해볼까요?

le2sky replied:
턴 개념을 추가해봤어요! 😀

seovalue commented on src/main/java/chess/view/command/MoveCommand.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    public static MoveCommand from(String input) {
+        if ("end".equals(input)) {
```
"end", " ", commandSegmentSize (3), "move"  이런 것들은 상수화해주시지 않은 이유가 궁금해요~!

le2sky replied:
상수로 추출한 것들(COMMAND_INDEX..)은 이름을 지어주지 않으면, 그것의 의미가 무엇인지 어려워서 상수로 추출했습니다.

move, end, start, 3, 2, " ", ""는 오히려, 같은 라인에 있어야 더욱 이해하기 쉽다고 느껴졌어요. 현재 move, end, start 키워드는 클래스 내부 코드에 중복이 많다고 느껴져 상수로 분리하도록 했습니다! 일관성있게 다른 것들도 상수로 바꾸는 것이 나을까요?

seovalue replied:
> move, end, start, 3, 2, " ", ""는 오히려, 같은 라인에 있어야 더욱 이해하기 쉽다고 느껴졌어요.

이 부분! 저도 공감해요ㅎㅎ 
해당 상수들을 반복적으로 사용하는 케이스가 생기면 그 때 분리해보죠~

seovalue commented on src/main/java/chess/view/command/MoveCommand.java:
```diff
@@ -0,0 +1,92 @@
+        return createDataFilledCommand(input);
+    }
+
+    private static MoveCommand createDataFilledCommand(String input) {
```
fillMoveCommand 이런식으로 변환되는 값을 메서드명에 나타내보는 건 어때요?

le2sky replied:
createFilleMoveCommand 로 메서드 이름을 변경해봤어요!

seovalue commented on src/main/java/chess/view/command/MoveCommand.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    public Coordinate source() {
+        validateEmptyData();
```
👍

seovalue commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -1,19 +1,14 @@
-    private static final int INITIAL_WHITE_PAWN_RANK = 2;
-    private static final int INITIAL_BLACK_SPECIAL_RANK = 8;
-    private static final int INITIAL_BLACK_PAWN_RANK = 7;
+    private static final Piece EMPTY_PIECE = new Empty();
```
empty piece에 대한 정의를 board에 해주신 이유가 궁금해요~

seovalue replied:
그리고 Empty 보다는 EmptyPiece라고 명명해주는 것은 어때요?

le2sky replied:
> empty piece에 대한 정의를 board에 해주신 이유가 궁금해요~

보드에서 `new Empty()`를 호출해주고, "음 이정도는 상수로 추출해볼까?"라는 의식의 흐름이 있었어요! 
현재는 Empty의 정적 팩터리 메서드를 사용하도록 변경했습니다.

> 그리고 Empty 보다는 EmptyPiece라고 명명해주는 것은 어때요?

Empty 보다 훨씬 명확한 이름 같아요! 감사합니다. 🙇🏻‍♂️

seovalue commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -1,19 +1,14 @@
-    private static final int INITIAL_BLACK_PAWN_RANK = 7;
+    private static final Piece EMPTY_PIECE = new Empty();
 
     private final Map<Coordinate, Piece> pieces = new HashMap<>();
```
pieces도 한번 감싸보는 건 어때요?

le2sky replied:
Pieces를 감싸도록 변경해봤습니다! 그리고 board에서 view나 piece 쪽으로 pieces를 내려주도록 변경했습니다!

`pieces.swap`의 경우에는 package-private으로 선언해 board의 외부에서 map을 변경하지 못하도록 구현해봤어요! 어색한 부분이 있다면, 추가 피드백 부탁드립니다. 🙇🏻‍♂️

seovalue commented on src/main/java/chess/domain/piece/AbstractSlidingPiece.java:
```diff
@@ -0,0 +1,58 @@
+import chess.domain.piece.exception.InvalidMoveException;
+import chess.domain.piece.exception.ObstacleException;
+
+abstract class AbstractSlidingPiece extends AbstractPiece {
```
SlidingPiece에서 의미하는 Sliding은 또 무슨 의미예요?

le2sky replied:
제가 용어 설명을 명확하게 해드리지 못해, 혼란을 드렸을 것 같아요. 죄송합니다. 🥲

> 현재는 다른 기물의 방해없이 쭉 미끄러질 수 있는 경로를 뜻하는 SlidingPath로 변경해봤어요.

위 코멘트에서 다른 기물의 방해없이 쭉 미끄러질 수 있는 경로를 Sliding Path라고 말씀 드렸는데요. 
(다른 기물의 방해가 있어도) 그냥 쭉 미끄러질 수 있는 경로를 Sliding Path라고 하며,  
이러한 경로를 이용해, 보드 위를 움직일 수 있는 녀석들(퀸, 비숍, 룩)을 Sliding Piece로 분류 해봤습니다!

seovalue replied:
@le2sky 제 생각에 sliding은 다른 사람이 보기에 한번에 이해하기 어려운 단어인 것 같아요.그리고 어떤 점에서 다른 기물의 방해없이 이동할 수 있다고 네이밍해주신 건지 궁금해요! 제가 알기론 퀸, 비숍, 룩은 앞에 다른 말이 있으면 통과하지 못하는 걸로 알아서요!

le2sky replied:
@seovalue 

제가 명확하게 설명드리지 못했네요. 🥲
단순하게, 칸 수 제한없이 이동할 수 있는 특성을 가진 기물을 Sliding Piece라고 생각했어요.
그리고, 칸 수 제한없이 이동하는 경로를 Sliding Path라고 생각했습니다.
그렇다고, 다른 기물의 방해없이 이동할 수 있다는 뜻은 아니었습니다. 
단지 칸 수 제한없이 이동한다에 초점을 맞춘 것입니다. 

> 제 생각에 sliding은 다른 사람이 보기에 한번에 이해하기 어려운 단어인 것 같아요.

더 좋은 단어를 찾아보려고 노력하겠습니다. 감사합니다. 🙇🏻‍♂️

seovalue commented on src/main/java/chess/domain/piece/AbstractSlidingPiece.java:
```diff
@@ -0,0 +1,58 @@
+
+    private List<Coordinate> createSlidingPath(Coordinate start, Direction direction) {
+        List<Coordinate> slidingPath = new ArrayList<>();
+        Weight weight = direction.getWeight();
```
뭔가 dx, dy에 대한 값을 weight라고 명명해주신 것 같은데, 가중치는 더해지는 값이라는 느낌이 드는 것 같아서요. 어떻게 생각하세요?!

le2sky replied:
확실히 가중치는 더해지는 값이 맞아요. 따라서, 문서와 테스트 코드 명에서 "더한다"라는 말을 사용하도록 수정해봤어요! 
coordinate.apply와 cooridinate.isApplicable 에 대해서 다시 좋은 이름을 지어주도록 노력해보겠습니다. 😀

seovalue commented on src/main/java/chess/domain/piece/exception/ObstacleException.java:
```diff
@@ -0,0 +1,10 @@
+package chess.domain.piece.exception;
+
+public class ObstacleException extends IllegalStateException {
```
Custom Exception으로 분리해주는 기준이 뭐예요?

le2sky replied:
저는 표준 예외와 메시지로 예외를 표현하는 것을 선호해요. 하지만, 한계를 느끼면 커스텀 예외를 고려합니다. 

현재는 아래와 같은 예외를 반복해서 사용하는 곳이 너무 많다고 느껴져서, 분리했습니다!
```java
new IllegalStateException("해당 기물은 주어진 좌표로 이동할 수 없습니다.");
new IllegalStateException("기물로 막혀있어 이동할 수 없습니다.");
```

seovalue replied:
표준 예외를 사용하는 것을 선호하시는 이유도 궁금해욧~

le2sky replied:
커스텀 예외를 사용하면, 다른 개발자가 그 커스텀 예외마저 새로 익혀야합니다. 이러한 것들도 비용이라고 생각해요. 
또한, 커스텀 예외를 정의하기 위해서 클래스를 새로 만들어야 하니 번거롭기도 합니다. 

제가 생각하는 실력있는 개발자의 특징 중 하나는 다른 개발자가 만든 코드를 적절하게 재사용하는 것입니다.
표준 예외는 충분히 잘 만들어져 있고, 널리 알려져 있습니다. 표준 예외로도 예외 상황을 잘 표현할 수 있다면, 재사용하는 것이 좋다고 생각해요. (사용하지 않으면, 만든 사람이 서운하기도 할 것 같아요.)

조앤은 평소에 개발하실 때, 어떤 것을 선호하시나요? 그리고, 현업에서는 어떤 방식으로 예외를 다루는지도 궁금합니다. 감사합니다. 🙇🏻‍♂️

seovalue reviewed (changes requested):
안녕하세요, 아톰!
엄청 많이 리팩터링해주셨네요ㅋㅋ 변경하시면서 든 질문 남겨주시면 더 좋을 것 같아요.
기존 코멘트에 대한 답변 및 추가로 코멘트를 남겨뒀는데 확인 후 리뷰 요청 주세요~

le2sky committed:
docs: 턴 기능을 추가하기 위한 기능목록 작성

le2sky committed:
feat: 체스 턴 기능 구현

le2sky committed:
refactor: slidingPath를 더욱 명료한 이름으로 개선

le2sky committed:
refactor: controller 이름 변경

le2sky committed:
refactor: 가중치 관련 테스트 명 개선

le2sky committed:
docs: 문서 수정

- 구현 기능을 체크했다.
- 가중치와 관련된 문장을 명확하게 개선했다.

le2sky committed:
refactor: 중첩 리스트를 반환하지 않도록 개선

le2sky committed:
refactor: getter 대신 메시지를 던지도록 개선

le2sky committed:
refactor: 반복되는 명령어 문자열을 상수로 추출

le2sky committed:
refactor: 완전한 MoveCommand를 생성하는 메서드 이름 개선

le2sky committed:
refactor: Empty 피스 선언을 Empty 내부로 이동

le2sky committed:
refactor: Empty를 더욱 명확한 이름으로 변경

le2sky committed:
refactor: coordinate apply 메서드 이름 개선

le2sky committed:
docs: pieces 포장을 위한 문서 수정

le2sky committed:
refactor: board의 책임을 일부 Pieces로 위임

le2sky committed:
refactor: 기물에서 Pieces를 사용하도록 변경

le2sky committed:
refactor: 미사용 메서드 제거

le2sky committed:
refactor: BoardFactory를 PiecesFactory로 변경

le2sky committed:
refactor: 기물 초기 위치 테스트를 PiecesTest로 이동

le2sky committed:
refactor: View에서 Pieces를 사용하도록 변경

le2sky committed:
chore: import 문 최적화

le2sky committed:
docs: 구현 기능 체크

le2sky committed:
refactor: 메서드 선언 순서 개선

le2sky commented:
조앤 안녕하세요! 아톰입니다.

좋은 피드백 남겨주셔서 감사합니다. 제가 조앤의 의도를 잘못 해석했거나, 코드에 어색한 부분이 있다면 추가 피드백 부탁 드립니다. 🙇🏻‍♂️

추가로 리팩터링 과정에 궁금한 점이 있어서 질문 드려요.

**[상황]**
이번 리팩터링 과정에서 기물이 존재하지 않는 예외 상황을 널 객체 패턴으로 숨겼어요. 
```java
Piece piece = pieces.findByCoordinate(...) // 없다면, Empty 피스를 반환 !!
```

그 과정에서 `domain.EmptyPiece`에 대한 `PieceType`을 `PiecType.Empty`로 새로 정의해줬습니다. 아래는 `EmptyPiece`에 대한 생성자입니다.
```java
private EmptyPiece() {
   super(PieceType.EMPTY, Team.EMPTY);
}
```

`PieceType`에 새로운 값이 정의됐고, 이로 인해서 런타임에 예외가 발생했습니다.
`view.PieceShapeSelector`는 기물의 타입을 받아서, 콘솔 뷰에서 출력될 모양을 결정하는데요. 
이곳에서 새로운 값을 대응하지 못해서 예외가 발생했습니다.

이런 상황에서 switch-expression을 이용한다면, 문제를 해결할 수 있다고 생각해요. 
하지만, 프로그래밍 요구사항에 switch/case를 사용하지 말라는 말이 있기 때문에 망설여졌습니다.

**[질문]**
프로그래밍 요구사항을 의식하면서 개발하려고 노력합니다. 하지만, 프로그래밍 요구사항을 깨는 것이 좋은(혹은 어쩔 수 없는..?) 상황에서는 지키지 않는 것이 좋을까요? 조앤의 의견이 궁금합니다.

seovalue commented on src/main/java/chess/view/PieceShapeSelector.java:
```diff
@@ -0,0 +1,32 @@
+        return Arrays.stream(PieceShapeSelector.values())
+                .filter(it -> it.pieceType == pieceType)
+                .findFirst()
+                .orElseThrow(() -> new NoSuchElementException("존재하지 않는 기물입니다."))
```
> PieceType에 새로운 값이 정의됐고, 이로 인해서 런타임에 예외가 발생했습니다.
> view.PieceShapeSelector는 기물의 타입을 받아서, 콘솔 뷰에서 출력될 모양을 결정하는데요.
> 이곳에서 새로운 값을 대응하지 못해서 예외가 발생했습니다.
> 
> 이런 상황에서 switch-expression을 이용한다면, 문제를 해결할 수 있다고 생각해요.
> 하지만, 프로그래밍 요구사항에 switch/case를 사용하지 말라는 말이 있기 때문에 망설여졌습니다.

일단 이 질문에 대해서 fit하게 먼저 답변을 드려보자면, 요구사항을 지키는 것은 일정을 지키는 것보다 더 중요하다고 생각해요. 지금은 사소한 구문에 대한 요구사항이지만, 나중에는 비즈니스 로직에 대한 요구사항이 될 수도 있기 때문이에요.ㅎㅎ 
그래서 우선은 최대한 지키면서 더 나은 방향이 어떤 것이 있을까?를 고민해보는 습관을 가져보시면 좋을 것 같아요.

그리고 이거랑은 별개로, 저는 EmptyPiece가 꼭 필요한 존재인가? 라는 것에 대해 의문을 던지고 싶어요ㅎㅎ 이 부분에 대해서도 다음 단계 진행하시면서 한번 고민해주세요!

le2sky replied:
답변 감사합니다. 제 성장에 도움이 됐습니다. 😀
EmptyPiece에 대해서도 다시 한번 고민해보겠습니다!

seovalue commented on docs/README.md:
```diff
@@ -28,15 +39,13 @@
-    - [x] 기물이 갈 수 있는 곳이라면, 보드를 업데이트한다.
-- [x] 체스판은 주어진 좌표에 해당하는 기물을 반환할 수 있다.
+    - [x] Black Pawn = (A - H)7
+- [x] 기들들은 주어진 좌표에 해당하는 기물을 반환할 수 있다.
```
매우 소소하지만 오타! `기물들은`

le2sky replied:
헉... 감사합니다!

seovalue commented on src/main/java/chess/ChessController.java:
```diff
@@ -9,12 +9,12 @@
 import chess.view.command.StartCommand;
 
-class Controller {
+class ChessController {
```
👍 사실 그냥 chessGame이 되어도 무방하다고 생각하긴 해요ㅎㅎ

seovalue commented on src/main/java/chess/domain/board/Pieces.java:
```diff
@@ -0,0 +1,28 @@
+import chess.domain.piece.EmptyPiece;
+import chess.domain.piece.Piece;
+
+public class Pieces {
```
👍 
Pieces를 일급컬렉션화 하니깐 어떠셨나요? 
오히려 불편하던가요, 아니면 더 낫던가요?

le2sky replied:
확실히 더 나아졌다고 생각해요. 😀

기존 Board가 가지고 있는 책임을 Pieces에게 분산하면서, 유지보수하기 더 좋아진 것 같아요. 
또한, Board를 Piece에게 내려주면서, "아 설마, 다른 사람이 Piece에서 Board.move를 다시 호출하지 않겠지..?" 라는 생각을 가지고 있었어요. Pieces를 내려주도록 변경하면서, 조회 메서드만 Piece에서 접근할 수 있도록 개선되어 좋았습니다.

그리고, 무엇보다 재미있었습니다.
Board.findByCoordinate와 Board.isPiecePresent는 많은 곳에서 사용하고 있었어요.
이것을 Pieces로 이동 시키기 위해서 조심스럽게 일련의 계획을 세우고, 적용해야 했어요. 
이 과정이 마치 게임을 하는 것 같아서 좋았어요. 감사합니다. 🙇🏻‍♂️

seovalue commented on src/main/java/chess/domain/piece/AbstractNonSlidingPiece.java:
```diff
@@ -26,8 +26,8 @@ void validatePieceMoveRule(Coordinate source, Coordinate target, Board board) {
                 .map(Direction::getWeight)
-                .filter(source::isApplicable)
-                .map(source::apply)
+                .filter(source::canPlus)
+                .map(source::plus)
```
오 뭔가 canPlus 한 것들만 filter해서 plus 한다는게 꽤 직관적으로 이해돼요!

seovalue commented on src/test/java/chess/domain/board/PiecesTest.java:
```diff
@@ -0,0 +1,235 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class PiecesTest {
```
👍

seovalue reviewed (approved):
아톰~ 리팩터링 잘 수행해주셨네요!
전반적으로 용어에 대한 것만 다음 단계 진행하기 전에 좀 더 고민해주시면 좋을 것 같아요!
기존 코멘트에 대한 답변 및 소소한 코멘트 남겨뒀는데 한번 확인해주세요!
